examples/vue/persister/src/persister.ts

```diff
@@ -0,0 +1,128 @@
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+import { hashQueryKey } from '@tanstack/query-core'
+
+export type Promisable<T> = T | PromiseLike<T>
+
+export interface PersistedQuery {
+  buster: string
+  timestamp: number
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export type PersistRetryer = (props: {
+  persistedQuery: PersistedQuery
+  error: Error
+  errorCount: number
+}) => Promisable<PersistedQuery | undefined>
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<void>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions<QC extends QueryClient> {
+  // TODO: if we decide to move this to an API layer, we could make this work without passing queryClient
+  /**
+   * Query Client instance
+   */
+  queryClient: QC
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (client: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   */
+  maxAge?: number
+}
+
+export function createPersister<T, QC extends QueryClient>(
+  queryFn: () => Promise<T>,
+  {
+    queryClient,
+    storage,
+    buster = '',
+    maxAge = 1000 * 60 * 60 * 24,
+    serialize = JSON.stringify,
+    deserialize = JSON.parse,
+  }: StoragePersisterOptions<QC>,
+) {
+  return async (context: QueryFunctionContext) => {
+    const queryHash = hashQueryKey(context.queryKey)
+    const queryState = queryClient.getQueryState(context.queryKey)
+
+    if (!queryState?.data && storage != null) {
+      const storedData = await storage.getItem(queryHash)
+      if (storedData) {
+        const persistedQuery = deserialize(storedData) as PersistedQuery
+
+        if (persistedQuery.timestamp) {
+          const expired = Date.now() - persistedQuery.timestamp > maxAge
+          const busted = persistedQuery.buster !== buster
+          if (expired || busted) {
+            await storage.removeItem(queryHash)
+          } else {
+            queryClient.getQueryCache().build(
+              queryClient,
+              {
+                queryKey: context.queryKey,
+                queryHash: queryHash,
+              },
+              persistedQuery.state,
+            )
+            return Promise.resolve(persistedQuery.state.data as T)
```

Comment:
If you do a return here... wouldn't that mean if data is in cache it will never be refreshed breaking `stale-while-revalidate` use-cases?

Reply:
this branch is only executed if we don't have data yet, see [this line](https://github.com/TanStack/query/pull/5125/files#diff-2a40d1fa40a7ae31330acd1ba80cfde7a697273dd9ae0d159bcba7c91e5a3959R79). It's basically a way to lazily seed the cache

Reply:
A good question here is if we should
- `restore` and still `fetch from server`
- `restore` and `wait for next fetch trigger`

Cause i think in this case we would just restore during initial component mount and skip `queryFn` which is not entirely correct as we would like to get latest data while revalidating. So instead of return `setQueryData` in here?

Reply:
> restore and still fetch from server
restore and wait for next fetch trigger

this should depend on the fact if the data that was restored is still considered `fresh` or not - depending on `staleTime`. If the data is considered outdated - call the `queryFn` immediately (but still return stale data). If it's fresh, just return it ...